meadow-client/src/MainFrame.purs

```diff
@@ -547,38 +556,39 @@ evalF (ApplyTransaction next) = do
     Nothing -> pure next
 
 evalF (NextBlock next) = do
-  modifying (_marloweState <<< _blockNum) (add one)
+  saveInitialState
+  modifying _marloweState (addToNEL (updateStateP <<< (over _blockNum (add one))))
   updateState
   pure next
 
 evalF (AddAnyInput {person, anyInput} next) = do
-  modifying (_marloweState <<< _transaction <<< _inputs) ((flip snoc) anyInput)
+  modifying (_marloweState <<< _Head <<< _transaction <<< _inputs) ((flip snoc) anyInput)
```

Comment:
Looking at these, I think there's a very strong case for creating something along the lines of:

```
modifyingCurrentHistory l = modifying (_marloweState <<< _Head <<< l)
extendHistory f = modifying _marloweState (addToNEL f)
replaceHistory v = modifying (_marloweState) (const $ NEL.singleton v)
```

ie. I think these repeating patterns could be turned into something named, and that'll make it easier to see that you're doing the sensible thing in each place.

Reply:
Or at the very least:

```
_currentMarloweState = _marloweState <<< _Head
```

Reply:
ok, I've gone for
```
_currentMarloweState :: forall a. Lens' { marloweState :: NonEmptyList MarloweState | a } MarloweState
_currentMarloweState = _marloweState <<< _Head

_currentContract :: forall a. Lens'  { marloweState :: NonEmptyList MarloweState | a } (Maybe Contract)
_currentContract = _currentMarloweState <<< _contract

_currentTransaction :: forall a. Lens'  { marloweState :: NonEmptyList MarloweState | a } TransactionData
_currentTransaction = _currentMarloweState <<< _transaction

_currentInput  :: forall a. Lens'  { marloweState :: NonEmptyList MarloweState | a } InputData
_currentInput = _currentMarloweState <<< _input
```
which neatens things up quite a bit and makes sense in use. Not sure if the location in MainFrame.purs is correct though.

Reply:
Cool. I'd gravitate towards putting things like this in `Types.purs`...

Reply:
done.